### ISSUE TITLE
Mixtures (here: HFC-410A)

### ISSUE THREAD

**Issue by mikapfl** [openscm/openscm-units#10]
Hi,

I'm using your excellent openscm-units library, and noticed that some countries are reporting HFC-410A emissions, but openscm doesn't know about HFC-410A. At first, I was confused because [Wikipedia](https://en.wikipedia.org/wiki/R-410A) claims a GWP for HFC-410A, but I couldn't find it in any of the GWP tables; however, I finally realized that this is because HFC-410A is a mixture of HFC-32 and HFC-125. So, most countries report HFC-32 and HFC-125 individually and everything is fine using openscm-units, but some countries report HFC-410A separately, which openscm-units can't handle at the moment. Now I'm wondering if it makes sense to include HFC-410A in openscm-units or how to deal with mixtures more generally.

Ideas I had so far:
* Just very pragmatically include HFC-410A like any other gas with a combined GWP (Wikipedia claims a GWP of 2088). This would allow conversions into CO2, but no automatic conversion into HFC-32 and HFC-125.
* Don't deal with this in openscm-units, let users split into HFC-32 and HFC-125 themselves.
* Somehow shoehorn pint into parsing HFC-410A as a sum of two different units. Normally, a product of two different units is no problem, but a sum of two different units is rather atypical. I haven't figured out a way how to do that with pint, yet. Maybe using the same methods that are used for temperatures?

I'm happy to write a pull request, but wanted to see if you have any other, better idea how to handle well-known mixtures of gases.

Cheers,

Mika

**Comment by znicholls** [comment]
Hmmm tricky one. 

> * Just very pragmatically include HFC-410A like any other gas with a combined GWP (Wikipedia claims a GWP of 2088).

This sounds like a good starting point.

How we could then help users to split HFC-410A into HFC32 and HFC125 is then a second question I would say.

> * This would allow conversions into CO2, but no automatic conversion into HFC-32 and HFC-125

Is there a defined split e.g. is 1 HFC-410A equal to 0.6 HFC32 and 2 HFC125 or is this unknown? If it's a known split, we could do some tricks using contexts to help people get the ratios without too many headaches.

**Comment by mikapfl** [comment]
> Is there a defined split e.g. is 1 HFC-410A equal to 0.6 HFC32 and 2 HFC125 or is this unknown? If it's a known split, we could do some tricks using contexts to help people get the conversion factors without too many headaches.

Yes, `1 HFC410A` is usually considered exactly `0.5 HFC32` and `0.5 HFC125`, which is also where the Wikipedia GWP comes from (simply the average of the AR4 GWPs).

**Comment by mikapfl** [comment]
Hm, there seem to be a lot of mixture gases in the refrigerant space, [Wikipedia](https://en.wikipedia.org/wiki/Refrigerant) has even a section about notable blends. Do you think they are all in scope for openscm-units? If yes, I might write a patch to include them all and generate their GWPs from the constituents automatically.

**Comment by mikapfl** [comment]
Alternatively, a function to easily add a mixture "unit" and GWP to a unit_registry could be useful. That would give users the ability to add mixtures as needed without adding the maintenance burden of including every mixture engineers come up with to openscm-units.

**Comment by znicholls** [comment]
> Do you think they are all in scope for openscm-units?

I would say yes

> Alternatively, a function to easily add a mixture "unit" and GWP to a unit_registry could be useful.

We could add the following method to the public API (i.e. remove the underscore, test it and document it)? That should let people do everything they need?

https://github.com/openscm/openscm-units/blob/f6867998bff91cb8c48dc709bfa27857e94bffa3/src/openscm_units/unit_registry.py#L399

**Comment by mikapfl** [comment]
> I would say yes

Cool, I'll work on a pull request for them. Currently, I think it would be best to add something like `data/mixtures.json` or so where the parts are given and then add the units and conversions automatically from the values of the constituents. That way, they'll automatically get updated GWP values when `data/metric_conversions.csv` is updated. I have some other things on my plate, so I don't know if I'll finish that this week (I would also need to find a nice citable source).
Or would you prefer explicit definitions in `data/metric_conversions.csv` like I'm doing it for HFC410a atm?

> We could add the following method to the public API (i.e. remove the underscore, test it and document it)? That should let people do everything they need?

That looks great. If the need arises, a more specialized interface to define mixture units and conversions in one step could be nice, but it's not really needed, I think.

**Comment by znicholls** [comment]
> I think it would be best to add something like `data/mixtures.json` or so where the parts are given and then add the units and conversions automatically from the values of the constituents.

This sounds great.

> Or would you prefer explicit definitions in `data/metric_conversions.csv` like I'm doing it for HFC410a atm?

No I think you make a nice point that they would quickly fall over/get out of date.

> I don't know if I'll finish that this week (I would also need to find a nice citable source)

No rush from our side. We can also throw a warning if the source isn't citable/we're unsure, totally up to you. 

> That looks great.

Cool, do you want to have a crack at doing that (I can also do it, just might be slower).

**Comment by znicholls** [comment]
Closed by #16
